File: solutions/go/allergies/1/allergies.go
package allergies

import (
    "slices"
)

func pow2(n int) uint {
    if n == 0 {
        return 1
    }
    return 2 << (n-1)
}

var allergens = []string{
    "eggs", "peanuts", "shellfish", "strawberries", "tomatoes",
    "chocolate", "pollen", "cats"}

func Allergies(allergies uint) []string {
    // Ignores allergens with higher scores (256, 512, ...)
	allergies = allergies % pow2(len(allergens))
	var res []string
    value := pow2(len(allergens)-1)
    for i := len(allergens)-1; i >= 0; i-- {
        if allergies >= value {
            res = append(res, allergens[i])
            allergies -= value
        }
        value /= 2
    }
    return res
}

func AllergicTo(allergies uint, allergen string) bool {
    return slices.Contains(Allergies(allergies), allergen)
}

File: solutions/lua/allergies/1/allergies.lua
--[[
  Lua doesn't have bitwise and operators so we can't use the obviosu solution using masks.
  We can go through allergies from highest to lowest score:
    - if the score is higher or equal then the allergy is in the list
    - if it is in the list, we need to remove it before checking the next one.
]]

local ALLERGIES = {
  "cats",
  "pollen",
  "chocolate",
  "tomatoes",
  "strawberries",
  "shellfish",
  "peanuts",
  "eggs",
}

local function list(score)
  -- since we don't have proper mask, limits the values to the first 8 bits
  score = score % 256
  local allergy_score = math.floor(2 ^ (#ALLERGIES - 1))
  local res = {}
  for _, allergie in ipairs(ALLERGIES) do
    if math.floor(score / allergy_score) == 1 then
      table.insert(res, 1, allergie)
      score = score - allergy_score
    end
    allergy_score = math.floor(allergy_score / 2)
  end
  return res
end

local function allergic_to(score, which)
  local allergies = list(score)
  for _, allergie in ipairs(allergies) do
    if allergie == which then return true end
  end
  return false
end

return {
  list = list,
  allergic_to = allergic_to
}
